proposals/extensions.md

```diff
@@ -1219,34 +1219,68 @@ implicit extension E for int
-The marker method is called `<ImplicitExtension>$` for implicit extensions and 
+It is private and static, and is called `<ImplicitExtension>$` for implicit extensions and 
 `<ExplicitExtension>$` for explicit extensions.  
+
 For example: `implicit extension R for UnderlyingType : BaseExtension1, BaseExtension2` yields
 `private static void <ImplicitExtension>$(UnderlyingType, BaseExtension1, BaseExtension2)`.  
 
-If the extension has any instance member, then we emit a ref field (of underlying type)
-into the ref struct and a constructor.  
-TODO2(instance) The wrapping can be done with a static unspeakable factory method  
+## Instance field
 
-Values of extension types are left as values of the underlying value type, until an extension
-member is accessed. When an extension member is accessed, an extension instance is created
-with a reference to the underlying value and the member is accessed on that instance.
+If the extension not static, then we emit a private instance field of the underlying type
+into the struct.  
+We use the default layout for structs (sequential layout, with pack and size 0)
+ensuring that the single instance field is placed at offset zero.
 
+Note: Although we could not find an explicit statement to that effect,
+this behavior falls out from ECMA 335 (II.10.1.2), which states for sequential layout:
+> The CLI shall lay out the fields in sequential order, based on the order of the fields in the logical metadata table (§II.22.15).
+
+> [Rationale: ... sequential layout is intended to instruct the CLI to match layout rules commonly followed by languages like C and C++
+> on an individual platform, where this is possible while still guaranteeing verifiable layout. ...]
+
+and from the C99 standard section 6.7.2.1 bullet point 13 (gated):
+
+> Within a structure object, the non-bit-field members and the units in which bit-fields 
+> reside have addresses that increase in the order in which they are declared. 
+> A pointer to a structure object, suitably converted, points to its initial member 
+> (or if that member is a bit-field, then to the unit in which it resides), 
+> and vice versa. There may be unnamed padding within a structure object, but not at its beginning.
+
+For example `implicit extension E for UnderlyingType` yields
+```csharp
+struct E
+{
+	private UnderlyingType <UnderlyingInstance>$;
+	private static void <ImplicitExtension>$(UnderlyingType) { }
+}
 ```
-Extension r = default(UnderlyingType); // emitted as a local of type `UnderlyingType`
-r.ExtensionMember(); // emitted as `new Extension(ref r).ExtensionMember();`
-```
+
+## Instance invocations
+
+The design and layout of the instance field will allow us to re-interpret an instance of `UnderlyingType`
+as an instance of `E` with `public static ref TTo Unsafe.As<TFrom,TTo>(ref TFrom source)`.  
+
+If the receiver of an extension member invocation on a value of the underlying type is `r`,
+we will replace it with `Unsafe.As<UnderlyingType, E>(ref r)`.
```

Comment:
This is not going to work in trimming or Native AOT. If the IL never contains a reference to the private field of the lowered struct, then it will trim it away. This will break the layout and the runtimes will crash on invocation.

Reply:
Actually, even with the marker we can't do anything for trimming/AOT in .NET 8. If someone compiles with extensions for netstandard2.0, and someone attempts to trim code with this pattern, they will hit code corruption and crash.

Reply:
It should be able to work if it generates something like this in private implementation details, right?
```csharp
static class Helper<T1, T2>
{
    [DynamicDependency(.NonPublicFields, typeof(T2))] static ref T2 AsExtension(ref T1 ut) => ref Unsafe.As<T1, T2>(ref r);
    [DynamicDependency(.NonPublicFields, typeof(T1))] static ref T2 AsUnderlying(ref T1 ut) => ref Unsafe.As<T1, T2>(ref r);
}
```